subprojects/core/src/main/java/org/gradle/internal/service/scopes/DefaultGradleUserHomeScopeServiceRegistry.java

```diff
@@ -117,25 +117,22 @@ public Optional<ServiceRegistry> getCurrentServices() {
     public void release(ServiceRegistry registry) {
         lock.lock();
         try {
-            for (Map.Entry<File, Services> entry : servicesForHomeDir.entrySet()) {
-                Services services = entry.getValue();
-                if (services.registry == registry) {
-                    if (services.count <= 0) {
-                        break;
-                    }
-                    services.count--;
-                    if (services.count == 0 && (servicesForHomeDir.size() > 1 || System.getProperty(REUSE_USER_HOME_SERVICES, "true").equals("false"))) {
-                        // Other home dir in use, close these. Otherwise, keep the services for next time
-                        CompositeStoppable.stoppable(services.registry).stop();
-                        servicesForHomeDir.remove(entry.getKey());
+            servicesForHomeDir.entrySet().stream()
+                .filter(entry -> entry.getValue().registry == registry && entry.getValue().count > 0)
+                .findFirst()
+                .ifPresent(
+                    entry -> {
+                        Services services = entry.getValue();
+                        services.count--;
+                        if (services.count == 0 && (servicesForHomeDir.size() > 1 || System.getProperty(REUSE_USER_HOME_SERVICES, "true").equals("false"))) {
+                            CompositeStoppable.stoppable(services.registry).stop();
+                            servicesForHomeDir.remove(entry.getKey());
+                        }
                     }
-                    return;
-                }
-            }
+                );
         } finally {
             lock.unlock();
         }
-        throw new IllegalStateException("Gradle user home directory scoped services have already been released.");
```

Comment:
Is this still needed? Feels a little off, only working in combination with the `return` in `L:132`.

Since loop is an actual, this might not be mandatory anymore.